src/openrct2/scenario/Scenario.cpp

```diff
@@ -427,10 +427,15 @@ static int32_t scenario_create_ducks()
     int32_t i, j, r, c, x, y, waterZ, centreWaterZ, x2, y2;
 
     // Originally, this function generated coordinates from 64 to 191.
-    // It now generates coordinates from 0 to 255, so smaller maps may also spawn ducks.
+    // The duck create function is called with 2 tiles of randomness to the target coordinate.
+    // The duck create function itself adds approximately 1 tile of randomness.
+    // To spawn ducks safely, the generated coordinates should not be in the outer 3 tiles of the map.
+
     r = scenario_rand();
-    x = ((r >> 16) & 0xFFFF) % MAXIMUM_MAP_SIZE_TECHNICAL;
-    y = (r & 0xFFFF) % MAXIMUM_MAP_SIZE_TECHNICAL;
+    x = 1 + DUCK_TARGET_TILES_AWAY_FROM_EDGE
```

Comment:
That expression got even more complex, either use scenario_rand_max or create a function that returns two ranged numbers from the initial random value. It's really hard to grasp whats happening here on a first look.

Reply:
A random 32 bit int r is generated.
For the x coordinates, it right shifts the number by 16 bits, so the leftmost 16 bits of r are used. 
For the y coordinate, the number is masked with 0xFFFF so the rightmost 16 bits of r are used.

Reply:
It is pointless to bitmask first and then modulo. Basically, you're modulo'ing it twice.